src/backend/processor/diagnostics.ts

```diff
@@ -0,0 +1,204 @@
+import { Event, EventEmitter, ExtensionContext, TextEditor, Uri, window } from 'vscode';
+import { parseDiagnostics } from '../parser';
+import { CheckerMetadata, DiagnosticEntry, DiagnosticFile } from '../types';
+import { ExtensionApi } from '../api';
+
+/**
+ * API interface that provides Diagnostics data.  
+ * Access the active instance via ExtensionApi.
+ */
+export class DiagnosticsApi {
+    private _openedFiles: string[] = [];
+
+    /** 
+     * Content based on the contents of loaded .plist files.
+     * Key: Source code file, value: .plist analysis files
+     */
+    private _diagnosticSourceFiles: Map<string, string[]> = new Map();
+
+    /**
+     * Loaded diagnostic files.
+     * Key: .plist analysis file, value: file's contents
+     */
+    private _diagnosticEntries: Map<string, DiagnosticFile> = new Map();
+
+    private _selectedEntry?: {
+        file: string,
+        idx: number
+    };
+    public get selectedEntry(): {
+        position: {readonly file: string, readonly idx: number},
+        diagnostic: DiagnosticEntry
+    } | undefined {
+        if (this._selectedEntry) {
+            const activeFile = this.getFileDiagnostics(Uri.file(this._selectedEntry.file)) ?? [];
+            const diagnostic = activeFile[this._selectedEntry.idx];
+
+            if (diagnostic) {
+                return {
+                    position: this._selectedEntry!,
+                    diagnostic
+                };
+            }
+        }
+
+        return undefined;
+    }
+
+    public setSelectedEntry(position?: {file: string, idx: number}) {
+        if (position) {
+            const activeFile = this.getFileDiagnostics(Uri.file(position.file)) ?? [];
+            const diagnostic = activeFile[position.idx];
+
+            if (diagnostic) {
+                this._selectedEntry = position;
+                this._diagnosticsUpdated.fire();
+                return;
+            }
+        }
+
+        this._selectedEntry = undefined;
+        this._diagnosticsUpdated.fire();
+    }
+
+    constructor(ctx: ExtensionContext) {
+        ctx.subscriptions.push(this._diagnosticsUpdated = new EventEmitter());
+        window.onDidChangeVisibleTextEditors(this.onDocumentsChanged, this, ctx.subscriptions);
+        ExtensionApi.metadata.metadataUpdated(this.onMetadataUpdated, this, ctx.subscriptions);
+
+        this.init();
+    }
+
+    private init() {
+        // Initialize the currently opened windows
+        this.onDocumentsChanged(window.visibleTextEditors);
+    }
+
+    // Logs errors to the user, and always resolves.
+    async reloadDiagnosticsSafe(forceReload?: boolean): Promise<void> {
+        try {
```

Comment:
The regular try/catch is not able to catch exception from an async function without await:
```ts
async function foo(): Promise<void> {
  return new Promise((res, rej) => {
    rej("Failure");
  })
}

async function fooSafe(): Promise<void> {
  try {
    foo();
  } catch (err) {
    console.error("fooSafe", err);
  }
}

async function fooSafe2(): Promise<void> {
  try {
    await foo();
  } catch (err) {
    console.error("fooSafe2", err);
  }
}

async function fooSafe3(): Promise<void> {
  foo().then(() => {
    console.log("SUCCESS");
  }).catch(err => {
    console.error("fooSafe3", err);
  });
}

// This will not catch the error.
fooSafe();

// Using await. This will catch the error.
fooSafe2();

// This will catch the error too.
fooSafe3();
```

Reply:
Thank you, fixed.